pkg/replicas/getter.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// the code below implements the integration of dispersed replicas in chunk fetching.
+// using storage.Getter interface.
+package replicas
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/file/redundancy"
+	"github.com/ethersphere/bee/pkg/soc"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+// ErrSwarmageddon is returned in case of a vis mayor called Swarmageddon.
+// Swarmageddon is the situation when none of the replicas can be retrieved.
+// If 2^{depth} replicas were uploaded and they all have valid postage stamps
+// then the probability of Swarmageddon is less than 0.000001
+// assuming the error rate of chunk retrievals stays below the level expressed
+// as depth by the publisher.
+type ErrSwarmageddon struct {
+	error
+}
+
+func (err *ErrSwarmageddon) Unwrap() []error {
+	if err == nil || err.error == nil {
+		return nil
+	}
+	var uwe interface{ Unwrap() []error }
+	if !errors.As(err.error, &uwe) {
+		return nil
+	}
+	return uwe.Unwrap()
+}
+
+// getter is the private implementation of storage.Getter, an interface for
+// retrieving chunks. This getter embeds the original simple chunk getter and extends it
+// to a multiplexed variant that fetches chunks with replicas.
+//
+// the strategy to retrieve a chunk that has replicas can be configured with a few parameters:
+//   - RetryInterval: the delay before a new batch of replicas is fetched.
+//   - depth: 2^{depth} is the total number of additional replicas that have been uploaded
+//     (by default, it is assumed to be 4, ie. total of 16)
+//   - (not implemented) pivot: replicas with address in the proximity of pivot will be tried first
+type getter struct {
+	wg sync.WaitGroup
+	storage.Getter
+	level redundancy.Level
+}
+
+// NewGetter is the getter constructor
+func NewGetter(g storage.Getter, level redundancy.Level) storage.Getter {
+	return &getter{Getter: g, level: level}
+}
+
+// Get makes the getter satisfy the storage.Getter interface
+func (g *getter) Get(ctx context.Context, addr swarm.Address) (ch swarm.Chunk, err error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// channel that the results (retrieved chunks) are gathered to from concurrent
+	// workers each fetching a replica
+	resultC := make(chan swarm.Chunk)
+	// errc collects the errors
+	errc := make(chan error, 17)
+	var errs []error
```

Comment:
Not need to be a slice, the `errors.Join` will handle the wrapping. Also, it should be declared as close to the for loop as possible.

Reply:
Still unresolved.